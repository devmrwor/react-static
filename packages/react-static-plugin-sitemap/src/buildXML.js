import fs from 'fs-extra'
import nodePath from 'path'
import { pathJoin } from 'react-static'
import chalk from 'chalk'

const REGEX_TO_GET_LAST_SLASH = /\/{1,}$/gm

const defaultGetUrlAttributes = (route, { prefixPath }) => {
  const { sitemap: { loc, ...rest } = {} } = route
  return {
    loc: getPermaLink(loc || route.path, prefixPath),
    ...rest,
  }
}

export default async function main(state, options) {
  const {
    config: {
      paths: { DIST },
      disableRoutePrefixing,
      siteRoot,
    },
    staging,
  } = state

  const prefixPath = disableRoutePrefixing
    ? siteRoot
    : process.env.REACT_STATIC_PUBLIC_PATH

  const filename = staging ? 'sitemap.staging.xml' : 'sitemap.xml'

  console.log(`Generating ${filename}...`)

  const xml = generateXML(state, options, prefixPath)
  await fs.writeFile(nodePath.join(DIST, filename), xml)

  console.log(chalk.green(`[\u2713] ${filename} generated`))
}

export function generateXML(
  state,
  { getAttributes = () => ({}) } = {},
  prefixPath
) {
  const { config, routes, staging } = state

  if (!prefixPath) {
    throw new Error('The sitemap url prefix cannot be empty or undefined!')
  }

  const xmlRoutes = routes
    .filter(r => {
      // Don't include the 404 page
      if (r.path === '404') {
        return false
      }
      // Don't include routes with noindex: true
      if (r.noindex) {
        return false
      }
      return true
    }) // Leave out noindex routes
    .map(route => {
      const attributes = {
        ...defaultGetUrlAttributes(route, {
          config,
          prefixPath,
        }),
        ...getAttributes(route, {
          config,
          prefixPath,
        }),
        noindex: undefined,
      }
      const attributesArr = []
      Object.keys(attributes).forEach(key => {
        if (typeof attributes[key] !== 'undefined') {
          attributesArr.push({ key, value: attributes[key] })
        }
      })
      return [
        '<url>',
        ...attributesArr.map(
          ({ key, value }) => `<${key}>${encode(value)}</${key}>`
        ),
        '</url>',
      ].join(staging ? '\n' : '')
    })
    .join(staging ? '\n' : '')

  return [
    `<?xml version="1.0" encoding="UTF-8"?>`,
    `<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">`,
    xmlRoutes,
    `</urlset>`,
  ].join(staging ? '\n' : '')
}

export function getPermaLink(path, prefixPath) {
  const permalink = `${prefixPath}${pathJoin(path)}`
  return `${permalink}/`.replace(REGEX_TO_GET_LAST_SLASH, '/')
}

function encode(val) {
  return String(val).replace(/[<>&'"]/g, c => {
    switch (c) {
      case '<':
        return '&lt;'
      case '>':
        return '&gt;'
      case '&':
        return '&amp;'
      case "'":
        return '&apos;'
      case '"':
        return '&quot;'
      default:
        throw new Error('XML encoding failed')
    }
  })
}
