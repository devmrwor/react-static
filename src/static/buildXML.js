import fs from 'fs-extra'
import path from 'path'

import { pathJoin } from '../utils/shared'

const REGEX_TO_GET_LAST_SLASH = /\/{1,}$/gm

export const getPermaLink = ({ path, prefixPath }) => {
  const permalink = `${prefixPath}${pathJoin(path)}`
  return `${permalink}/`.replace(REGEX_TO_GET_LAST_SLASH, '/')
}

export const makeGenerateRouteXML = ({ prefixPath }) => route => {
  const { path, lastModified, priority = 0.5 } = route
  return [
    '<url>',
    `<loc>${getPermaLink({ path, prefixPath }).replace(/[<>&'"]/g, c => {
      switch (c) {
        case '<':
          return '&lt;'
        case '>':
          return '&gt;'
        case '&':
          return '&amp;'
        case "'":
          return '&apos;'
        case '"':
          return '&quot;'
        default:
          throw new Error('XML encoding failed')
      }
    })}</loc>`,
    lastModified ? `<lastmod>${lastModified}</lastmod>` : '',
    `<priority>${priority}</priority>`,
    '</url>',
  ].join('')
}

export const generateXML = ({ routes, prefixPath }) =>
  `<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">${routes
    .filter(r => r.path !== '404')
    .filter(r => !r.noindex)
    .map(makeGenerateRouteXML({ prefixPath }))
    .join('')}</urlset>`

export const getSiteRoot = ({ stagingSiteRoot, siteRoot }) =>
  process.env.REACT_STATIC_STAGING === 'true' ? stagingSiteRoot : siteRoot

export default async ({ config }) => {
  const {
    routes, publicPath, paths = {}, disableRoutePrefixing,
  } = config

  const { DIST } = paths
  const siteRoot = getSiteRoot(config)
  const prefixPath = disableRoutePrefixing ? siteRoot : publicPath

  if (!siteRoot) {
    return
  }

  const xml = generateXML({ routes, prefixPath })

  await fs.writeFile(path.join(DIST, 'sitemap.xml'), xml)
}
