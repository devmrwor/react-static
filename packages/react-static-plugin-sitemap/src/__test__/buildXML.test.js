import fsExtra from 'fs-extra'
import { getPermaLink, generateXML } from '../buildXML'

describe('getPermaLink', () => {
  it('should return a permalink', () => {
    const permalink = getPermaLink('/path/to/somewhere', '/blog/')
    expect(permalink).toEqual('/blog/path/to/somewhere/')
  })
})

describe('generateXML', () => {
  it('should return an xml string', () => {
    const xml = generateXML(
      {
        routes: [
          {
            path: '/path/to/somewhere',
          },
        ],
      },
      undefined,
      '/blog/'
    )

    expect(typeof xml).toEqual('string')
    expect(xml).toMatchSnapshot()
  })

  it('should require a siteRoot', () => {
    expect(() =>
      generateXML({
        routes: [
          {
            path: '/path/to/somewhere',
          },
        ],
      })
    ).toThrow()
  })

  it('should utilize custom sitemap properties', () => {
    const xml = generateXML(
      {
        routes: [
          {
            path: '/path/to/somewhere',
            sitemap: {
              lastmod: '10/10/2010',
              priority: '0.5',
              'image:image': {
                'image:loc': `https://raw.githubusercontent.com/react-static/react-static/master/media/react-static-logo-2x.png`,
                'image:caption': 'React Static',
              },
            },
          },
        ],
      },
      undefined,
      '/blog/'
    )

    expect(xml).toMatchSnapshot()
  })

  it('should use encoding for XML-values', () => {
    const xml = generateXML(
      {
        routes: [
          {
            path: '/"官话"-is-chinese-\'ру́сский язы́к\'-is-russian',
          },
        ],
      },
      undefined,
      '/this-&-that/'
    )

    expect(xml).toMatchSnapshot()
  })

  it('should return a the XML without noindex and 404 routes ', () => {
    const xml = generateXML(
      {
        routes: [
          { path: '/path/to/article/' },
          { path: '/path/to/somewhere/', noindex: true },
          { path: '404' },
        ],
      },
      undefined,
      '/blog/'
    )

    expect(xml.split('<loc>').length).toEqual(2)
  })

  it('should support a custom getAttributes option ', () => {
    const xml = generateXML(
      {
        routes: [
          {
            path: '/path/to/somewhere',
          },
        ],
      },
      {
        getAttributes: () => ({
          test: 'foobar',
        }),
      },
      '/blog/'
    )

    expect(xml.includes('<test>foobar</test>')).toEqual(true)
  })
})
