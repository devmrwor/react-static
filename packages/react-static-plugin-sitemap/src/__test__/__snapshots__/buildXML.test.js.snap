// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateXML should return an xml string 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\"><url><loc>/blog/path/to/somewhere/</loc></url></urlset>"`;

exports[`generateXML should use encoding for XML-values 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\"><url><loc>/this-&amp;-that/&quot;官话&quot;-is-chinese-&apos;ру́сский язы́к&apos;-is-russian/</loc></url></urlset>"`;

exports[`generateXML should utilize custom sitemap properties 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\"><url><loc>/blog/path/to/somewhere/</loc><lastmod>10/10/2010</lastmod><priority>0.5</priority></url></urlset>"`;
